### DISCUSSION TITLE
How to implement custom post processing effects into Filament?

### DISCUSSION THREAD

**Discussion by Nolram12345** [google/filament#7676] [Q&A]
Hi there!
For a current project, we are looking to implement 2 custom post processing effects into filament. These are just simple full-screen effects, specifically a light shaft ("godrays") effect, and chromatic aberration.

Is there a framework or other methodology for implementing custom post processing effects into filament?

**Comment by romainguy** [comment]
It's currently not directly possible. You can do it via off screen render targets but it would be best to be able to plug into the frame graph.

@pixelflinger

**Reply by Nolram12345** [reply]
Hm, that's unfortunate. Are there any plans for potentially adding at least a simple interface that would allow adding simple full-screen single-shader custom post processing effects? Or is there perhaps some other workaround that we could use?

**Reply by pixelflinger** [reply]
We have had such plans since we implemented the FrameGraph, unfortunately it's low on our priority list.

**Reply by Nolram12345** [reply]
Are there any other ways to implement a custom post processing effect into Filament? Is it possible to do in a at least somewhat straight-forward way with a custom filament build? What is needed to register a new effect with Filament?

**Reply by Nolram12345** [reply]
@romainguy You mentioned it's possible via off screen render targets. Could you elaborate a bit on that?

**Reply by poweifeng** [reply]
You can check out this example for an idea of how it works: https://github.com/google/filament/blob/main/samples/rendertarget.cpp

**Reply by Nolram12345** [reply]
Had a look at it - so essentially the process would be to render the whole scene into a rendertarget, applying a custom material for the desired post-processing effect, and then blitting/rendering that to the screen again? That seems quite inefficient for essentially just running a full-screen shader over the output - I think it would be much more stable and straight forward for me to fork filament and just add a new effect internally :/ 

I really hope there will be a post processing API in the future - something that just allows one to sample the rendered scene color & maybe depth, run that through a custom material and have that output to the screen would be amazing.

**Reply by kihaki** [reply]
Hey guys, is there any update on this? Is it possible to apply fullscreen custom shader effects to the output?
